Timeout automatic status updates

Parveshdhull opened the pull request:
fixes https://github.com/status-im/status-mobile/issues/13601

### Summary
Currently, we are handling timing out of automatic status updates after 5 minutes on the client-side. PR moves this functionality to status-go.

Now we will receive `status.updates.timedout` signals once the status is timed out.

**status-react PR**: https://github.com/status-im/status-mobile/pull/13602

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?
- [ ] Have you tested changes with mobile?
- [x] Have you tested changes with desktop?

Parveshdhull commented on protocol/messenger_communities.go:
```diff
@@ -1314,13 +1314,13 @@ func (m *Messenger) passStoredCommunityInfoToSignalHandler(communityID string) {
-	if community.Name() == "" && community.DescriptionText() == "" && community.MembersCount() == 0 {
+	if err != nil {
+		m.logger.Warn("cant get community and pass it to signal handler", zap.Error(err))
 		return
 	}
 
-	if err != nil {
-		m.logger.Warn("cant get community and pass it to signal handler", zap.Error(err))
+	//if there is no info helpful for client, we don't post it
+	if community.Name() == "" && community.DescriptionText() == "" && community.MembersCount() == 0 {
```
Change is unrelated to PR, but makes sure that we check error before accessing community.

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (30)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | d0bdbd04 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/1/) | 2022-07-18 21:58:18 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220718-215549-d0bdbd-pr2757.zip) |
| :heavy_check_mark: | d0bdbd04 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/1/) | 2022-07-18 22:00:28 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220718-215549-d0bdbd-pr2757.aar) |
| :heavy_check_mark: | d0bdbd04 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/1/) | 2022-07-18 22:01:15 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220718-215549-d0bdbd-pr2757.zip) |
| | | | | | | |
| :heavy_check_mark: | 250c1b62 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/2/) | 2022-07-21 21:50:21 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220721-214731-250c1b-pr2757.zip) |
| :heavy_check_mark: | 250c1b62 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/2/) | 2022-07-21 21:53:00 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220721-214727-250c1b-pr2757.aar) |
| :heavy_check_mark: | 250c1b62 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/2/) | 2022-07-21 21:53:26 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220721-214727-250c1b-pr2757.zip) |
| | | | | | | |
| :heavy_check_mark: | 9bcc475c | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/3/) | 2022-07-21 22:03:03 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220721-220057-9bcc47-pr2757.zip) |
| :heavy_check_mark: | 9bcc475c | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/3/) | 2022-07-21 22:03:25 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220721-220057-9bcc47-pr2757.zip) |
| :heavy_check_mark: | 9bcc475c | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/3/) | 2022-07-21 22:06:01 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220721-220057-9bcc47-pr2757.aar) |
| | | | | | | |
| :heavy_check_mark: | c4cfd3be | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/4/) | 2022-07-21 22:59:53 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220721-225758-c4cfd3-pr2757.zip) |
| :heavy_check_mark: | c4cfd3be | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/4/) | 2022-07-21 23:00:28 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220721-225756-c4cfd3-pr2757.zip) |
| :heavy_check_mark: | c4cfd3be | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/4/) | 2022-07-21 23:01:21 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220721-225756-c4cfd3-pr2757.aar) |
| | | | | | | |
| :heavy_check_mark: | 3ff1a6f2 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/5/) | 2022-07-27 09:46:32 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220727-094420-3ff1a6-pr2757.zip) |
| :heavy_check_mark: | 3ff1a6f2 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/5/) | 2022-07-27 09:47:34 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220727-094420-3ff1a6-pr2757.zip) |
| :heavy_check_mark: | 3ff1a6f2 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/5/) | 2022-07-27 09:48:32 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220727-094420-3ff1a6-pr2757.aar) |
| | | | | | | |
| :heavy_check_mark: | 828a7f34 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/6/) | 2022-07-27 11:27:00 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220727-112509-828a7f-pr2757.zip) |
| :heavy_check_mark: | 828a7f34 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/6/) | 2022-07-27 11:27:42 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220727-112509-828a7f-pr2757.zip) |
| :heavy_check_mark: | 828a7f34 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/6/) | 2022-07-27 11:28:46 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220727-112507-828a7f-pr2757.aar) |
| | | | | | | |
| :heavy_check_mark: | 59d0e8fe | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/7/) | 2022-07-29 01:27:02 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220729-012506-59d0e8-pr2757.zip) |
| :heavy_check_mark: | 59d0e8fe | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/7/) | 2022-07-29 01:27:51 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220729-012506-59d0e8-pr2757.zip) |
| :heavy_check_mark: | 59d0e8fe | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/7/) | 2022-07-29 01:28:32 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220729-012504-59d0e8-pr2757.aar) |
| | | | | | | |
| :heavy_check_mark: | 2440cec8 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/8/) | 2022-07-29 01:32:47 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220729-013046-2440ce-pr2757.zip) |
| :heavy_check_mark: | 2440cec8 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/8/) | 2022-07-29 01:33:09 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220729-013042-2440ce-pr2757.zip) |
| :heavy_check_mark: | 2440cec8 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/8/) | 2022-07-29 01:33:54 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220729-013042-2440ce-pr2757.aar) |
| | | | | | | |
| :heavy_check_mark: | f1ee85b7 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/9/) | 2022-07-29 02:06:52 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220729-020516-f1ee85-pr2757.zip) |
| :heavy_check_mark: | f1ee85b7 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/9/) | 2022-07-29 02:07:47 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220729-020516-f1ee85-pr2757.zip) |
| :heavy_check_mark: | f1ee85b7 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/9/) | 2022-07-29 02:08:33 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220729-020514-f1ee85-pr2757.aar) |
| | | | | | | |
| :heavy_check_mark: | 4fe632d2 | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/10/) | 2022-07-29 02:09:38 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220729-020807-4fe632-pr2757.zip) |
| :heavy_check_mark: | 4fe632d2 | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/10/) | 2022-07-29 02:10:21 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220729-020802-4fe632-pr2757.zip) |
| :heavy_check_mark: | 4fe632d2 | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/10/) | 2022-07-29 02:11:39 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220729-020836-4fe632-pr2757.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 23ed753e | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/11/) | 2022-07-29 08:46:03 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220729-084421-23ed75-pr2757.zip) |
| :heavy_check_mark: | 23ed753e | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/11/) | 2022-07-29 08:47:06 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220729-084421-23ed75-pr2757.zip) |
| :heavy_check_mark: | 23ed753e | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/11/) | 2022-07-29 08:47:39 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220729-084419-23ed75-pr2757.aar) |
| | | | | | | |
| :heavy_check_mark: | 086a0f72 | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2757/12/) | 2022-07-29 08:52:56 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220729-085134-086a0f-pr2757.zip) |
| :heavy_check_mark: | 086a0f72 | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2757/12/) | 2022-07-29 08:54:00 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220729-085134-086a0f-pr2757.zip) |
| :heavy_check_mark: | 086a0f72 | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2757/12/) | 2022-07-29 08:54:42 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220729-085132-086a0f-pr2757.aar) |

Samyoul commented on protocol/messenger_status_updates.go:
```diff
@@ -267,3 +267,63 @@ func (m *Messenger) HandleStatusUpdate(state *ReceivedMessageState, statusMessag
+	// but the range covers special cases like, other status updates had the same clock value
+	// or the received another status update with higher clock value than the reference clock but
+	// lower clock value than the statusUpdate.Clock
+	deactivatedStatusUpdates, err := m.persistence.DeactivatedStatusUpdates(fromClock, tillClock)
```
I think that if `persistence` returns an error it would be useful to know what that error is. Otherwise this function will just silently fail

Samyoul commented on protocol/messenger_status_updates.go:
```diff
@@ -267,3 +267,63 @@ func (m *Messenger) HandleStatusUpdate(state *ReceivedMessageState, statusMessag
+	if err == nil {
+		// Extra 5 sec wait (broadcast receiving delay)
+		waitDuration = statusUpdate.Clock + fiveMinutes + 5 - now
+	} else {
```
Maybe add a `DEBUG` level log on the error, just to catch unexpected errors

Samyoul reviewed (commented):
Fundamentally this a sound PR. I've made a couple of comments about error handling.

Your `timeoutAutomaticStatusUpdates` func should probably have a test to validate its behaviour and the 2 `LastStatusUpdateFromClock` and `DeactivatedStatusUpdates` funcs should have tests to confirm their behaviours

Parveshdhull commented:
hi @Samyoul, Thank you very much for reviewing PR. New Changes:
- Renamed `LastStatusUpdateFromClock` to `NextHigherClockValueOfAutomaticStatusUpdates`, because I am only using clock value from status update, so no point in returning whole status update.
- Added tests for `DeactivatedAutomaticStatusUpdates` and `NextHigherClockValueOfAutomaticStatusUpdates` as you requested.
- Added error debug logs and removed some duplicate code
- Behaviour of `timeoutAutomaticStatusUpdates` tested in [status-react PR](https://github.com/status-im/status-mobile/pull/13602), but not sure how I write unit test for this one. Any ideas?

endulab commented on protocol/messenger_config.go:
```diff
@@ -44,6 +44,7 @@ type MessengerSignalsHandler interface {
 	HistoryArchivesSeeding(communityID string)
 	HistoryArchivesUnseeded(communityID string)
 	HistoryArchiveDownloaded(communityID string, from int, to int)
+	StatusUpdatesTimedOut(statusUpdates interface{})
```
Why you use an empty interface instead of a concrete type (UserStatus) ?

Parveshdhull replied:
Replaced with `*[]protocol.UserStatus`

endulab commented on protocol/messenger_status_updates.go:
```diff
@@ -267,3 +267,59 @@ func (m *Messenger) HandleStatusUpdate(state *ReceivedMessageState, statusMessag
+	if err == nil {
+		m.config.messengerSignalsHandler.StatusUpdatesTimedOut(&deactivatedStatusUpdates)
+	} else {
+		m.logger.Debug("Unable to deactivate atutomatic status updates", zap.Error(err))
```
atutomatic  -> automatic, rather "Unable to fetch/get deact...." because you read from db and not deactivate, right ?

Parveshdhull replied:
Fixed typo and debug log

endulab reviewed (changes requested):
@Parveshdhull what about the 2 week period mentioned in specs ? Is it handled somehow ?

Parveshdhull commented:
Hi @endulab, Thank you very much for reviewing PR.
> [@Parveshdhull](https://github.com/Parveshdhull) what about the 2 week period mentioned in specs ? Is it handled somehow ?

PR only implements timing out of automatic status updates so that we don't need a running timer in the client. 
For `always online` status updates(2 weeks), we don't need a timer. We only check once while loading status updates if the status update is 2 weeks old or not.
Also, we continuously delete status updates older than 2 weeks, so this check is also not much necessary.
https://github.com/status-im/status-go/blob/6d38b860598e2442e1c4256f9445455a2fcb9d8c/protocol/persistence.go#L967

Parveshdhull committed:
Timeout automatic status updates

Parveshdhull commented:
**Testing Request:**
Please check if status updates, (especially automatic status updates) are working as expected and nothing is broken.

elina2015 reviewed (approved):
Tested and approved on the Desktop app
